### PR TITLE
Fixed #24361 -- Clarified docs on reconfiguring logging.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/24361

### CODE REVIEW

File: docs/topics/logging.txt [hunk 1]
```diff
@@ -219,9 +219,17 @@ configuration merged with Django's defaults, hence you can decide if you want to
...
 If the ``disable_existing_loggers`` key in the :setting:`LOGGING` dictConfig is
-set to ``True`` (which is the default) the default configuration is completely
-overridden. Alternatively you can redefine some or all of the loggers by
-setting ``disable_existing_loggers`` to ``False``.
+set to ``True`` (which is the default) then all loggers from the default
+configuration will be disabled. Alternatively you can redefine some or all of
+the loggers by setting ``disable_existing_loggers`` to ``False``.
+
+.. note::
+    There is an unintuitive aspect to how Python handles disabled loggers.
```
Comment (berkerpeksag) [inline docs/topics/logging.txt]: Perhaps we can improve stdlib logging documentation and even the API too (for Python 3.5 only)?

**Comment by phalt** [comment]
:+1: looks good.

File: docs/topics/logging.txt [hunk 2]
```diff
@@ -219,9 +219,17 @@ configuration merged with Django's defaults, hence you can decide if you want to
...
 
 If the ``disable_existing_loggers`` key in the :setting:`LOGGING` dictConfig is
-set to ``True`` (which is the default) the default configuration is completely
-overridden. Alternatively you can redefine some or all of the loggers by
-setting ``disable_existing_loggers`` to ``False``.
+set to ``True`` (which is the default) then all loggers from the default
+configuration will be disabled. Alternatively you can redefine some or all of
+the loggers by setting ``disable_existing_loggers`` to ``False``.
+
+.. note::
```
Comment (carljm) [inline docs/topics/logging.txt]: I think this information should be worked directly into the above paragraph, not pulled out in a note.

File: docs/topics/logging.txt [hunk 3]
```diff
@@ -219,9 +219,17 @@ configuration merged with Django's defaults, hence you can decide if you want to
 add to, or replace the existing configuration.
 
 If the ``disable_existing_loggers`` key in the :setting:`LOGGING` dictConfig is
-set to ``True`` (which is the default) the default configuration is completely
-overridden. Alternatively you can redefine some or all of the loggers by
-setting ``disable_existing_loggers`` to ``False``.
+set to ``True`` (which is the default) then all loggers from the default
+configuration will be disabled. Alternatively you can redefine some or all of
```
Comment (carljm) [inline docs/topics/logging.txt]: "...will be disabled. This is _not_ the same as removed; the logger will still exist, but will silently discard anything logged to it, not even propagating it to a parent logger. Thus you should be very careful using `disable_existing_loggers: True`; it's probably not what you want. Instead, you can set `disable_existing_loggers` to `False` and redefine some or all of the default loggers; or you can set `LOGGING_CONFIG` to `None` and handle logging config yourself."

File: docs/topics/logging.txt [hunk 4]
```diff
@@ -398,12 +408,43 @@ configure logging using your own approach), you can set
...
+process is disabled, not logging itself. If you disable the configuration
+process, Django will still make logging calls, falling back to whatever default
+logging behavior is defined.
+
+Here's an example of how to make the logging system print everything to the
+console without having Django's disabled default loggers swallow entries. It
+may be useful during local development. Set the environment variable
+``DJANGO_LOG_LEVEL=DEBUG`` to see all of Django's logging calls in the console.
+
+.. snippet::
```
Comment (carljm) [inline docs/topics/logging.txt]: This example may be useful to include, but it does not belong in this section. You can use this config to get useful console logging while leaving Django's default logging config also in place, so use of this configuration is orthogonal to the question of setting `LOGGING_CONFIG` to `None` to disable Django's default logging config.

File: docs/topics/logging.txt [hunk 5]
```diff
@@ -398,12 +408,43 @@ configure logging using your own approach), you can set
...
-    configuration process is disabled, not logging itself. If you
-    disable the configuration process, Django will still make logging
-    calls, falling back to whatever default logging behavior is
-    defined.
+Setting :setting:`LOGGING_CONFIG` to ``None`` only means that the configuration
+process is disabled, not logging itself. If you disable the configuration
+process, Django will still make logging calls, falling back to whatever default
+logging behavior is defined.
+
+Here's an example of how to make the logging system print everything to the
```
Comment (carljm) [inline docs/topics/logging.txt]: The only example that should be included here is showing how to import logging and configure it yourself (if you want to bypass Django's default logging config); showing a sample alternate logging config is off-topic for this section.

We should also link to the docs for Django's default config here, so you know what configuration you're bypassing by using this approach.

File: docs/topics/logging.txt [hunk 6]
```diff
@@ -398,12 +408,43 @@ configure logging using your own approach), you can set
...
+    LOGGING = {
+        'version': 1,
+        'handlers': {
+            'console': {
+                'level': os.getenv('DJANGO_LOG_LEVEL', 'INFO'),
+                'class': 'logging.StreamHandler',
+                'formatter': 'debug',
+            },
+        },
+        'formatters': {
```
Comment (carljm) [inline docs/topics/logging.txt]: This isn't really necessary if we're aiming for a minimal example; it's a nicety, but the default formatter is reasonable too.

File: docs/topics/logging.txt (hunk 4 above)
Comment (timgraham) [inline docs/topics/logging.txt]: How can you get messages from the `django.request` logger while leaving Django's default configuration in place? Here is what it does:
```
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': False,
        },
```
Here is what I have in my test project:
```
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'console': {
            'level': os.getenv('LOGLEVEL', 'DEBUG'),
            'class': 'logging.StreamHandler',
            'formatter': 'debugging',
        },
    },
    'formatters': {
        'debugging': {
            'format': '%(name)s %(levelname)s %(module)s:%(lineno)s: %(message)s'
        },
    },
    'root': {
        'handlers': ['console'],
        'level': 'DEBUG',
    },
    'django.request': {
        'handlers': ['console'],
        'level': 'DEBUG',
        'propagate': True,
    },
}
```

File: docs/topics/logging.txt (hunk 6 above)
Comment (timgraham) [inline docs/topics/logging.txt]: okay

File: docs/topics/logging.txt [hunk 7]
```diff
@@ -390,20 +418,30 @@ use a different configuration process, you can use any other callable
...
+
+.. snippet::
+    :filename: settings.py
+
+    LOGGING_CONFIG = None
+
+    import logging.config
+    logging.config.dictConfig(...)
+
+Setting :setting:`LOGGING_CONFIG` to ``None`` only means that the configuration
```
Comment (carljm) [inline docs/topics/logging.txt]: "the _automatic_ configuration process"?

File: docs/topics/logging.txt [hunk 8]
```diff
@@ -261,7 +267,29 @@ First, here's a simple configuration which writes all request logging from the
...
+        'root': {
+            'handlers': ['console'],
+            'level': 'DEBUG',
+        },
+    }
+
+Finally, here's an example of a fairly complex logging setup, configured using
 :func:`logging.config.dictConfig`::
 
     LOGGING = {
```
Comment (carljm) [inline docs/topics/logging.txt]: Hmm, this example also uses `disable_existing_loggers: True`, which is bad. I don't think we should show any examples in our docs with `disable_existing_loggers`. This example should perhaps also be changed to set `LOGGING_CONFIG` to `None` and use `dictConfig` directly in order to get the desired effect? (Although in this case since the default logger for `django.request` is redefined here, maybe that re-enables it anyway?)

File: docs/topics/logging.txt [hunk 9]
```diff
@@ -261,7 +267,29 @@ First, here's a simple configuration which writes all request logging from the
 If you use this example, be sure to change the ``'filename'`` path to a
 location that's writable by the user that's running the Django application.
 
-Second, here's an example of a fairly complex logging setup, configured using
+Second, here's an example of how to make the logging system print everything to
+the console based on the value of an environment variable . It may be useful
+during local development. Set the environment variable
+``DJANGO_LOG_LEVEL=DEBUG`` to see all of Django's logging calls in the console::
+
+    # Not working yet. Entries from django.request and django.security aren't
```
Comment (carljm) [inline docs/topics/logging.txt]: Yeah, you're right, I was wrong that this works fully with the default config in place. Sigh. I don't like having all our logging examples bypass Django's logging config system entirely, but perhaps that's really our only accurate option until a proper fix. It seems like both the "console log in dev" example and the "complex" example really should bypass the default config.

It may also be that a "proper fix"  could be as simple as a new `LOGGING_SKIP_DEFAULTS` setting, which could be set to `True` to skip ever installing the default logging config? Then that setting would perform the function that these docs previously assumed `disable_existing_loggers` could perform, and it would simplify the boilerplate needed to do a full override with your own logging config. Instead of setting `LOGGING_CONFIG = None` and then having to import `logging` and call `logging.dictConfig` yourself, all you'd have to do is set `LOGGING_SKIP_DEFAULTS = True` and `LOGGING`.

(Or instead of defining a new setting, we could add our own custom key into `LOGGING`; but I don't like that as well, because it breaks the full equivalence between `LOGGING` and what `dictConfig` is documented to receive.)

I'll make this proposal on the mailing list.

**Comment by timgraham** [comment]
Carl, I updated the patch to also include more accurate info about the current state of loggers in Django. If this looks okay, I will commit and backport it, and then address improvements in a separate PR/ticket.

File: docs/topics/logging.txt [hunk 10]
```diff
@@ -261,12 +265,37 @@ First, here's a simple configuration which writes all request logging from the
 If you use this example, be sure to change the ``'filename'`` path to a
 location that's writable by the user that's running the Django application.
 
-Second, here's an example of a fairly complex logging setup, configured using
+Second, here's an example of how to make the logging system print all of
+Django's logging to the console based on the value of an environment variable.
+It overrides the fact that ``django.request`` and ``django.security`` don't
+propagate their log entries. It may be useful during local development. Set the
+environment variable ``DJANGO_LOG_LEVEL=DEBUG`` to see all of Django's logging
```
Comment (carljm) [inline docs/topics/logging.txt]: I'm not sure what the point of having the `DJANGO_LOG_LEVEL` environment variable in this example is; seems like needless added complexity that's beside the point of the example. I'd just remove the env var and set the level to `INFO`.

Also, the description of the effect of the env var is wrong here -- with the code as below, if you don't set the env var, you'll see anything of level `INFO` or higher. You can just set the env var to change that threshold level.

Comment (timgraham) [inline docs/topics/logging.txt]: The idea was just to suggest an easy way to toggle between DEBUG (very verbose, including all DB queries) and INFO (hardly anything) that didn't require modifying your settings file.

File: docs/topics/logging.txt [hunk 11]
```diff
@@ -261,12 +265,37 @@ First, here's a simple configuration which writes all request logging from the
...
+        },
+        'loggers': {
+            'django': {
+                'handlers': ['console'],
+                'level': os.getenv('DJANGO_LOG_LEVEL', 'INFO'),
+            },
+        },
+    }
+
+Finally, here's an example of a fairly complex logging setup, configured using
```
Comment (carljm) [inline docs/topics/logging.txt]: I know this wording pre-existed this PR, but it seems strange to me to say "configured using `logging.config.dictConfig`" here. All of these examples use `dictConfig`, because that's what `LOGGING_CONFIG` does by default. This example doesn't use `dictConfig` any more than anything in this section. And it seems to confuse with the technique below of setting `LOGGING_CONFIG=None` and then using `dictConfig` yourself directly.

I think it should be said somewhere in this section (if it isn't already) that the default behavior is to pass `settings.LOGGING` to `dictConfig` (and link to the `dictConfig` docs), but I'm not sure why it should be said specifically for this example.

File: docs/topics/logging.txt [hunk 12]
```diff
@@ -650,13 +687,24 @@ logging module.
...
 
-By default, Django configures the ``django.request`` logger so that all messages
-with ``ERROR`` or ``CRITICAL`` level are sent to :class:`AdminEmailHandler`, as
-long as the :setting:`DEBUG` setting is set to ``False``.
+By default, Django configures the following logging:
+
+When :setting:`DEBUG` is ``True``:
+
+* The ``django`` catch-all logger sends all messages at the ``INFO`` level or
+  higher to the console. Django doesn't make any such logging calls at this
```
Comment (carljm) [inline docs/topics/logging.txt]: I'm not sure what you're trying to say by "Django doesn't make any such logging calls at this time." Django does make quite a few logging calls that would be caught (via propagation) by the ``django`` catch-all logger. A quick grep turns up at least `django.contrib.gis` and various modules within `django.db`. It seems to be true that Django never logs anything directly to the `django` logger, but I don't think that's really relevant or worthy of mention; the point of the `django` catch-all is to catch things logged in sub-modules and propagated upwards.

Comment (carljm) [inline docs/topics/logging.txt]: Oh, I see - it's "at ``INFO`` level or higher" - Django's internal logging in those modules is at ``DEBUG`` level. I think that should be clarified.

File: docs/topics/logging.txt (hunk 10 above)
Comment (carljm) [inline docs/topics/logging.txt]: Ok - if we keep it in the example, I think it requires a bit more explanation. Something like "by default, this config only sends messages of level `INFO` or higher to the console, and Django itself does not currently log any such messages. Set the environment variable `DJANGO_LOG_LEVEL=DEBUG` to see all of Django's debug logging."

Comment (carljm) [inline docs/topics/logging.txt]: (I do see how the current wording is actually correct, because in practice Django only logs at `DEBUG` level. But I think it's unclear; at least it was to me.)

File: docs/topics/logging.txt [hunk 13]
```diff
@@ -261,12 +265,37 @@ First, here's a simple configuration which writes all request logging from the
 If you use this example, be sure to change the ``'filename'`` path to a
 location that's writable by the user that's running the Django application.
 
-Second, here's an example of a fairly complex logging setup, configured using
+Second, here's an example of how to make the logging system print all of
+Django's logging to the console based on the value of an environment variable.
+It overrides the fact that ``django.request`` and ``django.security`` don't
```
Comment (carljm) [inline docs/topics/logging.txt]: Wow - it does? I'll need to experiment and check that out. That doesn't seem to make any sense at all. Why would configuring a `django` logger override the previous configuration for the `django.request` and `django.security` loggers?

File: docs/topics/logging.txt [hunk 14]
```diff
@@ -650,13 +687,24 @@ logging module.
...
+* The ``py.warnings`` logger which handles messages from ``warnings.warn()``
+  sends messages to the console.
+
+When :setting:`DEBUG` is ``False``:
 
-All messages reaching the ``django`` catch-all logger when :setting:`DEBUG` is
-``True`` are sent to the console. They are simply discarded (sent to
-``NullHandler``) when :setting:`DEBUG` is ``False``.
+* The ``django.request`` and ``django.security`` loggers send messages with
+  ``ERROR`` or ``CRITICAL`` level are sent to :class:`AdminEmailHandler`. These
```
Comment (carljm) [inline docs/topics/logging.txt]: strike "are sent"

Comment (timgraham) [inline docs/topics/logging.txt]: done

File: docs/topics/logging.txt (hunk 12 above)
Comment (timgraham) [inline docs/topics/logging.txt]: propose to add: "(all logging is at the ``DEBUG`` level or handled by the
  ``django.request`` or ``django.security`` loggers)"

File: docs/topics/logging.txt (hunk 10 above)
Comment (carljm) [inline docs/topics/logging.txt]: I guess "Django itself does not currently log any such messages" isn't quite true, since it does within `django.security` and `django.request`.

File: docs/topics/logging.txt (hunk 13 above)
Comment (carljm) [inline docs/topics/logging.txt]: Wow - it does. Clearly I still don't quite understand the intended behavior of consecutive calls to `dictConfig` with `disable_existing_loggers: False`. I guess if you define a higher-level logger in a subsequent call, it replaces that entire portion of the logging "tree", replacing any previous configuration for sub-loggers of that logger.

File: docs/topics/logging.txt (hunk 11 above)
Comment (timgraham) [inline docs/topics/logging.txt]: removed "configured using dictConfig"

It's said above, "By default, Django uses the `dictConfig format`_." and below, "The LOGGING_CONFIG setting defines the callable that will be used to configure Django’s loggers. By default, it points at Python’s logging.config.dictConfig() function"

File: docs/topics/logging.txt (hunk 13 above)
Comment (carljm) [inline docs/topics/logging.txt]: I would maybe add "by default" here: "It overrides the default logging configuration, where ``django.request`` and ``django.security`` loggers don't propagate their log entries."

**Comment by carljm** [comment]
Modulo the above comments, this looks good to me as an incremental step, without making any implementation changes yet.

File: docs/topics/logging.txt (hunk 12 above)
Comment (carljm) [inline docs/topics/logging.txt]: perfect

File: docs/topics/logging.txt (hunk 13 above)
Comment (timgraham) [inline docs/topics/logging.txt]: done

File: docs/topics/logging.txt (hunk 10 above)
Comment (timgraham) [inline docs/topics/logging.txt]: changed "any such messages" to "many such messages"
